{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick and clean: Python for biological data processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [Course information](day1/info.ipynb)\n",
    "    - About the course\n",
    "    - About us\n",
    "    - About you\n",
    "    - Venue, credits, pricing, requirements\n",
    "    - Feedback\n",
    "    - Bibliography\n",
    "\n",
    "**Day 1, 8:00-16:00, Python**\n",
    "\n",
    "Easy introduction, light tasks.\n",
    "\n",
    "- [What is Python?](day1/what.ipynb)\n",
    "    - Stats, strengths and weaknesses\n",
    "    - Past, present and future of Python\n",
    "    - As a new \"pythonista\", who will be your friends?\n",
    "    - How to make Python work for this course\n",
    "    - Python distributions, Anaconda\n",
    "    - Jupyter and interactive notekeeping\n",
    "    - Installing libraries\n",
    "    - Python consoles, interpreters and editors\n",
    "- [Python tutorial](day1/tutorial.ipynb)\n",
    "    - Basics: Math, Variables, Functions, Control flow, Modules\n",
    "    - Data representation: String, Tuple, List, Set, Dictionary, Objects and Classes\n",
    "    - Standard library modules: script arguments, file operations, timing, processes, forks, multiprocessing\n",
    "- [Text manipulation](day1/text.ipynb):\n",
    "    - File IO, streaming, serialization\n",
    "    - Parsing and regular expressions\n",
    "    - XML, HTML editing\n",
    "- [Python and the web](day1/web.ipynb)\n",
    "    - Introduction to Django\n",
    "    - SQL interogation\n",
    "    - Remote API calls (Entrez, BioBank)\n",
    "- [Python and other languages](day1/languages.ipynb)\n",
    "    - Python and C: Mutual Information\n",
    "    - Python and R: microarray processing\n",
    "\n",
    "**Day 2, 8:00-16:00, Data science**\n",
    "\n",
    "Intensive in math, slightly harder tasks to accomplish in class.\n",
    "\n",
    "- [Visualization](day2/visualization.ipynb):\n",
    "    - Standard plots with matplotlib: line, scatter, chart\n",
    "    - Web publishing with plotly: heatmap example\n",
    "    - Network display with graphviz\n",
    "    - GUI programming with wxpython\n",
    "- [Statistics](day2/statistics.ipynb):\n",
    "    - Dataframing with pandas\n",
    "    - scipy: anova, linear regression, curve fitting\n",
    "    - Statistical enrichment analysis\n",
    "- [Scientific computing](day2/scicomp.ipynb):\n",
    "    - Numpy: advanced array operations\n",
    "    - Scipy introduction: from linear algebra to image analisys\n",
    "    - Simpy: symbolic math\n",
    "- [Machine learning](day2/learning.ipynb):\n",
    "    - scikit-learn: clustering\n",
    "    - Handling multivariate data: PCA and PLS regression\n",
    "- [Networks](day2/networks.ipynb):\n",
    "    - networkx: centrality computation\n",
    "    - Network IO\n",
    "- *Presentation of Omics*\n",
    "    - Omics tasks of day 3 are presented and discussed.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "**Day 3, 8:00-16:00, 'Omics**\n",
    "\n",
    "I setup the problems and describe the tasks, and give you some helper code to start with, and you will solve them in class, in the order of your choosing. I will tend to guide rather than tell.\n",
    "\n",
    "- [NGS_pipelining](day3/ngs_pipelining.ipynb):\n",
    "    - Open a cloud instance and install required programs\n",
    "    - Setup the pipeline\n",
    "    - Mapping and IGV inspection\n",
    "    - Normalizing counts and differential expression\n",
    "    - Galaxy integration\n",
    "- [Sequencing](day3/sequencing.ipynb)\n",
    "    - Make a toy sequencing library in standard Python for processing DNA, RNA and protein data.\n",
    "    - Implement the DNA, RNA and proteins as Python classes\n",
    "    - Make methods for transcription, translation, regulation.\n",
    "    - Compute several sequence similarity scores, such as hamming distance and mutual information.\n",
    "    - Add BioPython methods and prefix them with bp\n",
    "    - Describe your module in a tutorial like fashion\n",
    "- [Gene Expression](day3/expression.ipynb)\n",
    "    - Download a GEO dataset and prepare it\n",
    "    - Cluster the genes based on their expression\n",
    "    - Compute a co-expression network\n",
    "    - Compute differential gene expression for a set of samples.\n",
    "    - Compute functional enrichment of the main clusters.\n",
    "- [Transcriptomics](day3/transcriptomics.ipynb)\n",
    "    - Extract the promoter regions using biopython\n",
    "    - Investigate de-novo motifs on clusters of genes using meme or steme\n",
    "    - Use TransFac database to search for motif occurences on selected genes\n",
    "    - Save the found motifs and related data\n",
    "    - Test which of the motif occurences on your selected genes is significant\n",
    "- [Proteomics](day3/proteomics.ipynb)\n",
    "    - Compute a protein similarity graph, cluster enrichment study\n",
    "    - Perform structural alignment and plots with PyMol\n",
    "- [Metabolomics](day3/metabolomics.ipynb)\n",
    "    - Metabolic pathway assembly and display\n",
    "    - Flux balance analysis\n",
    "- [Dynamic modeling](day3/dynamic_modeling.ipynb)\n",
    "    - Load a curated SBML model\n",
    "    - Plot the model\n",
    "    - Solve the model\n",
    "    - Peak identification\n",
    "    - Pathway studies\n",
    "- [Population genetics and philogeny](PopGen.ipynb) - Removed/Not included.\n",
    "    - Run a small scale coalescent simulation\n",
    "    - Compute a philogeny tree and display it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
